### PR TITLE
[SPARK-29956][SQL] A literal number with an exponent should be parsed to Double

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

For a literal number with an exponent(e.g. 1e-45, 1E2), we'd parse it to Double by default rather than Decimal. And user could still use  `spark.sql.legacy.exponentLiteralToDecimal.enabled=true` to fall back to previous behavior. 

### Why are the changes needed?

According to ANSI standard of SQL, we see that the (part of) definition of `literal` :

```
<approximate numeric literal> ::=
    <mantissa> E <exponent> 
```
which indicates that a literal number with an exponent should be approximate numeric(e.g. Double) rather than exact numeric(e.g. Decimal). 

And when we test Presto, we found that Presto also conforms to this standard: 

```
presto:default> select typeof(1E2);
 _col0
--------
 double
(1 row)
```

```
presto:default> select typeof(1.2);
    _col0
--------------
 decimal(2,1)
(1 row)
```

We also find that, actually, literals like `1E2` are parsed as Double before Spark2.1, but changed to Decimal after #14828 due to *The difference between the two confuses most users* as it said. But we also see support(from DB2 test) of original behavior at #14828 (comment).

Although, we also see that PostgreSQL has its own implementation: 

```
postgres=# select pg_typeof(1E2);
 pg_typeof 
-----------
 numeric
(1 row)

postgres=# select pg_typeof(1.2);
 pg_typeof 
-----------
 numeric
(1 row)
```

We still think that Spark should also conform to this standard while considering SQL standard and Spark own history and majority DBMS and also user experience.

### Does this PR introduce any user-facing change?

Yes.

For `1E2`, before this PR:

```
scala> spark.sql("select 1E2")
res0: org.apache.spark.sql.DataFrame = [1E+2: decimal(1,-2)]
```

After this PR:

```
scala> spark.sql("select 1E2")
res0: org.apache.spark.sql.DataFrame = [100.0: double]
```

And for `1E-45`, before this PR:

```
org.apache.spark.sql.catalyst.parser.ParseException:
decimal can only support precision up to 38
== SQL ==
select 1E-45
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parse(ParseDriver.scala:131)
  at org.apache.spark.sql.execution.SparkSqlParser.parse(SparkSqlParser.scala:48)
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parsePlan(ParseDriver.scala:76)
  at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:605)
  at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
  at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:605)
  ... 47 elided
```

after this PR:

```
scala> spark.sql("select 1E-45");
res1: org.apache.spark.sql.DataFrame = [1.0E-45: double]
```

And before this PR, user may feel super weird to see that `select 1e40` works but `select 1e-40 fails`. And now, both of them work well.

### How was this patch tested?

updated `literals.sql.out` and `ansi/literals.sql.out`

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -23,6 +23,12 @@ grammar SqlBase;
    */
   public boolean legacy_setops_precedence_enbled = false;
 
+  /**
+   * When false, a literal with an exponent would be converted into
+   * double type rather than decimal type.
+   */
+  public boolean legacy_exponent_literal_to_decimal_enabled = false;
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: nit: `legacy_exponent_literal_as_decimal_enabled`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -1939,11 +1939,32 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
...
+  override def visitLegacyDecimalLiteral(ctx: LegacyDecimalLiteralContext)
+    : Literal = withOrigin(ctx) {
+    Literal(BigDecimal(ctx.getText).underlying())
+  }
+
+  /**
+   * Create a double literal for number with an exponent, e.g. 1E-30
+   */
+  override def visitExponentLiteral(ctx: ExponentLiteralContext): Literal = {
+    numericLiteral(ctx, Double.MinValue, Double.MaxValue, DoubleType.simpleString)(_.toDouble)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: shall we just call `ctx.getText.toDouble`?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: What `numericLiteral` does is to strip the suffix (which we don't have in expenent) and check range (is it really useful for double?)

File: sql/core/src/test/resources/sql-tests/results/ansi/literals.sql.out [hunk 3]
```diff
@@ -213,9 +219,15 @@ struct<0.3:decimal(1,1),-0.8:decimal(1,1),0.5:decimal(1,1),-0.18:decimal(2,2),0.
 -- !query 22
 select 123456789012345678901234567890123456789e10d, 123456789012345678901234567890123456789.1e10d
 -- !query 22 schema
-struct<1.2345678901234568E48:double,1.2345678901234568E48:double>
+struct<>
 -- !query 22 output
-1.2345678901234568E48	1.2345678901234568E48
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/literals.sql.out]: so the previous result was wrong?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 2 above)
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: For exponent values, we recognize it by searching "E" in `numericLiteral`, see:

```
val rawStrippedQualifier = if (text.toUpperCase(Locale.ROOT).contains("E")) {
  // exponent values don't have a suffix but contain 'E'
  text
} else {
  text.substring(0, ctx.getText.length - 1)
}
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: hmm this looks hacky. Can we move the strip of suffix to the caller side? then `numericLiteral` can focus on the range check.

Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Sounds good.

File: sql/core/src/test/resources/sql-tests/results/ansi/literals.sql.out (hunk 3 above)
Comment (Ngone51) [reply sql/core/src/test/resources/sql-tests/results/ansi/literals.sql.out]: No, I just realize that a bug has been introduced in `numericLiteral()` if we simply use `contains(E)` to recognize exponent value. Because we also have `DIGIT+ EXPONENT? 'D'`, e.g. 10E2D, for double, which results in collision.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 4]
```diff
@@ -1939,11 +1939,27 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
...
+  /**
+   * Create a double literal for number with an exponent, e.g. 1E-30
+   */
+  override def visitExponentLiteral(ctx: ExponentLiteralContext): Literal = {
+    numericLiteral(ctx, ctx.getText, /* exponent values don't have a suffix */
+      Double.MinValue, Double.MaxValue, DoubleType.simpleString)(_.toDouble)
+  }
+
   /** Create a numeric literal expression. */
   private def numericLiteral
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: nit: the code style should be
```
def func(
    para1: T,
    para2: T,
    ...): R = 
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -1937,6 +1937,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val LEGACY_EXPONENT_LITERAL_TO_DECIMAL_ENABLED =
+    buildConf("spark.sql.legacy.exponentLiteralToDecimal.enabled")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `exponentLiteralAsDecimal`

File: sql/core/src/test/resources/sql-tests/results/ansi/decimalArithmeticOperations.sql.out [hunk 6]
```diff
@@ -73,19 +73,17 @@ struct<(CAST(10.300000000000000000 AS DECIMAL(21,19)) * CAST(3.00000000000000000
 -- !query 8
 select (5e36 + 0.1) + 5e36
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/decimalArithmeticOperations.sql.out]: This test is to check the decimal overflow behavior. let's keep testing it: `(5e36BD + 0.1) + 5e36BD`

File: sql/core/src/test/resources/sql-tests/results/ansi/decimalArithmeticOperations.sql.out [hunk 7]
```diff
@@ -73,19 +73,17 @@ struct<(CAST(10.300000000000000000 AS DECIMAL(21,19)) * CAST(3.00000000000000000
...
-struct<>
+struct<((5.0E36 + CAST(0.1 AS DOUBLE)) + 5.0E36):double>
 -- !query 8 output
-java.lang.ArithmeticException
-Decimal(expanded,10000000000000000000000000000000000000.1,39,1}) cannot be represented as Decimal(38, 1).
+1.0E37
 
 
 -- !query 9
 select (-4e36 - 0.1) - 7e36
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/decimalArithmeticOperations.sql.out]: ditto

File: sql/core/src/test/resources/sql-tests/results/ansi/decimalArithmeticOperations.sql.out [hunk 8]
```diff
@@ -100,10 +98,9 @@ Decimal(expanded,1.5241578753238836750190519987501905210E+38,38,-1}) cannot be r
 -- !query 11
 select 1e35 / 0.1
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/decimalArithmeticOperations.sql.out]: ditto

File: sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-group-by.sql.out [hunk 9]
```diff
@@ -172,16 +172,16 @@ WHERE  t1b IN (SELECT Max(t2b)
                FROM   t2
                GROUP  BY t2a)
 -- !query 9 schema
-struct<t1a:string,t1b:smallint,t1c:int,t1d:bigint,t1e:float,t1f:double,t1g:decimal(2,-2),t1h:timestamp,t1i:date>
+struct<t1a:string,t1b:smallint,t1c:int,t1d:bigint,t1e:float,t1f:double,t1g:double,t1h:timestamp,t1i:date>
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-group-by.sql.out]: let's update the CREATE TABLE statements in this test files to create decimal column correctly. It tries to test both double and decimal type and we should keep the test coverage.

File: sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-limit.sql.out [hunk 10]
```diff
@@ -74,10 +74,10 @@ WHERE  t1a IN (SELECT t2a
                WHERE  t1d = t2d)
 LIMIT  2
 -- !query 3 schema
-struct<t1a:string,t1b:smallint,t1c:int,t1d:bigint,t1e:float,t1f:double,t1g:decimal(2,-2),t1h:timestamp,t1i:date>
+struct<t1a:string,t1b:smallint,t1c:int,t1d:bigint,t1e:float,t1f:double,t1g:double,t1h:timestamp,t1i:date>
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-limit.sql.out]: ditto

File: sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-order-by.sql.out [hunk 11]
```diff
@@ -73,13 +73,13 @@ WHERE  t1a IN (SELECT t2a
                FROM   t2)
 ORDER  BY t1a
 -- !query 3 schema
-struct<t1a:string,t1b:smallint,t1c:int,t1d:bigint,t1e:float,t1f:double,t1g:decimal(2,-2),t1h:timestamp,t1i:date>
+struct<t1a:string,t1b:smallint,t1c:int,t1d:bigint,t1e:float,t1f:double,t1g:double,t1h:timestamp,t1i:date>
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/subquery/in-subquery/in-order-by.sql.out]: ditto

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 12]
```diff
@@ -1939,11 +1939,29 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     Literal(BigDecimal(ctx.getText).underlying())
   }
 
+  /**
+   * Create a decimal literal for a regular decimal number.
+   */
+  override def visitLegacyDecimalLiteral(ctx: LegacyDecimalLiteralContext)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: nit:
```
def f(
    para: T): R
```

**Comment by cloud-fan** [comment]
also cc @viirya @maropu

**Comment by maropu** [comment]
Actually, it seems this behaviour depends on implementations (the standard says so though).
 In PgSQL;
```
postgres=# select pg_typeof(1E2);
 pg_typeof 
-----------
 numeric
(1 row)

postgres=# select pg_typeof(1.2);
 pg_typeof 
-----------
 numeric
(1 row)
```
Just in case, can you check the other system behaviour other than PgSQL/Presto?

btw, when the PgSQL dialect enabled, should we follow the current Spark behaviour (that is, in both cases, we regard them as `decimal`)?

**Comment by Ngone51** [comment]
Actually, literals like `1E2` are parsed as `Double` before Spark2.1, but changed to `Decimal` after #14828 due to "The difference between the two confuses most users" as it said. But I also see support(DB2 test) of original behavior at https://github.com/apache/spark/pull/14828#issuecomment-251805403.



> btw, when the PgSQL dialect enabled, should we follow the current Spark behaviour (that is, in both cases, we regard them as decimal)?

hmm... as for this, I'd say yes. But I agree that things here may be a little awkward when we switch to PostgreSQL....as we need to back to legacy Decimal mode again.

**Comment by maropu** [comment]
> Actually, literals like 1E2 are parsed as Double before Spark2.1, but changed to Decimal after #14828 due to "The difference between the two confuses most users" as it said. But I also see support(DB2 test) of original behavior at #14828 (comment).

Ah, I see. Thanks for that info. Either way, I personally think its worth adding a note about this behaivour change in the migration guide. Can you update?

> hmm... as for this, I'd say yes. But I agree that things here may be a little awkward when we switch to PostgreSQL....as we need to back to legacy Decimal mode again.

NVM, if necessary, we can update that behaivour in follow-up. Thanks, anyway.

**Comment by cloud-fan** [comment]
yea a migration guide is needed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 13]
```diff
@@ -1949,11 +1949,29 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     Literal(BigDecimal(ctx.getText).underlying())
   }
 
+  /**
+   * Create a decimal literal for a regular decimal number.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: this also covers exponent number, right?

**Comment by viirya** [comment]
Not sure if this is necessary. Looks like we have changed this behavior before. Previous reason is for user experience, and this is for standard?

**Comment by cloud-fan** [comment]
I'd say this change can improve UX. It's super weird that `select 1e40` works but `select 1e-40` fails

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 13 above)
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: I think you're right, though it's copied from original comment. I'll append some comments for exponent number too.

**Comment by Ngone51** [comment]
```
01:02:36.134 ERROR org.apache.spark.sql.SQLQueryTestSuite: Error using configs: 
[info] - postgreSQL/numeric.sql *** FAILED *** (1 minute, 14 seconds)
[info]   postgreSQL/numeric.sql
[info]   Expected "struct<LOG([CAST(1.2345678E-28 AS DOUBLE)]):double>", but got "struct<LOG([1.2345678E-28]):double>" Schema did not match for query #550
[info]   select ln(1.2345678e-28): QueryOutput(select ln(1.2345678e-28),struct<LOG(1.2345678E-28):double>,-64.26166165451762) (SQLQueryTestSuite.scala:419)
```

How does this query passed previously? It's quite weird.

File: docs/sql-migration-guide.md [hunk 14]
```diff
@@ -225,6 +225,8 @@ license: |
   - Since Spark 3.0, when casting string value to integral types, including tinyint, smallint, int and bigint type, the leading and trailing white spaces(<= ACSII 32) will be trimmed before convert to integral values, e.g. `cast(' 1 ' as int)` results `1`. In Spark version 2.4 and earlier, the result will be `null`.
 
   - Since Spark 3.0, when casting string value to date, timestamp and interval values, the leading and trailing white spaces(<= ACSII 32) will be trimmed before casing, e.g. `cast('2019-10-10\t as date)` results the date value `2019-10-10`. In Spark version 2.4 and earlier, only the trailing space will be removed, thus, the result is `null`.
+  
+  - Since Spark 3.0, the scientific decimal literal(e.g. `1E2`) would be parsed as Double while in Spark version 2.4 it's parsed as Decimal. To restore the behavior before Spark 3.0, you can set `spark.sql.legacy.exponentLiteralAsDecimal.enabled` to `true`.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: `the scientific decimal literal` -> `numbers written in scientific notation`

Comment (cloud-fan) [reply docs/sql-migration-guide.md]: And please follow the style of other migration guides
```
Since Spark 3.0, .... In Spark version 2.4 and earlier, ...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 15]
```diff
@@ -1943,6 +1943,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val LEGACY_EXPONENT_LITERAL_AS_DECIMAL_ENABLED =
+    buildConf("spark.sql.legacy.exponentLiteralAsDecimal.enabled")
+      .internal()
+      .doc("When set to true, a literal with an exponent (e.g. 1E-30) would be parsed" +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: nit: needs a space at the end

**Comment by cloud-fan** [comment]
LGTM

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala [hunk 16]
```diff
@@ -473,13 +473,15 @@ class ExpressionParserSuite extends AnalysisTest {
...
-    testDecimal("9.0e1")
-    testDecimal(".9e+2")
-    testDecimal("0.9e+2")
-    testDecimal("900e-1")
-    testDecimal("900.0E-1")
-    testDecimal("9.e+1")
+    // SPARK-29956: Scientific Decimal is parsed as Double by default.
+    assertEqual("9.0e1", Literal(90.toDouble))
+    assertEqual(".9e+2", Literal(90.toDouble))
+    assertEqual("0.9e+2", Literal(90.toDouble))
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: Can we add test case for legacy?

File: docs/sql-migration-guide.md [hunk 17]
```diff
@@ -225,6 +225,8 @@ license: |
   - Since Spark 3.0, when casting string value to integral types, including tinyint, smallint, int and bigint type, the leading and trailing white spaces(<= ACSII 32) will be trimmed before convert to integral values, e.g. `cast(' 1 ' as int)` results `1`. In Spark version 2.4 and earlier, the result will be `null`.
 
   - Since Spark 3.0, when casting string value to date, timestamp and interval values, the leading and trailing white spaces(<= ACSII 32) will be trimmed before casing, e.g. `cast('2019-10-10\t as date)` results the date value `2019-10-10`. In Spark version 2.4 and earlier, only the trailing space will be removed, thus, the result is `null`.
+  
+  - Since Spark 3.0, numbers written in scientific notation(e.g. `1E2`) would be parsed as Double. In Spark version 2.4 and earlier, they're parsed as Decimal. To restore the behavior before Spark 3.0, you can set `spark.sql.legacy.exponentLiteralAsDecimal.enabled` to `true`.
```
Comment (maropu) [inline docs/sql-migration-guide.md]: nit: `numbers` -> `numeric literals`?

**Review by maropu — APPROVED** [review]
LGTM except for one minor comment.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala (hunk 16 above)
Comment (Ngone51) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: Added ca56ceb

**Comment by dongjoon-hyun** [comment]
@Ngone51 . Could you summarize the previous discussions (on the above comments) into the PR description in terms of the reference PostgreSQL and old Spark 2.1 behavior?

Also, cc @gatorsmile

**Comment by Ngone51** [comment]
@dongjoon-hyun OK.

File: docs/sql-migration-guide.md (hunk 17 above)
Comment (Ngone51) [reply docs/sql-migration-guide.md]: Wait, wait. `numeric` represents **exact** numeric/number in pg sql. But we may wish an approximate numeric/number here as we want to parse it as Double. So, in order to not make user be confused with these concepts, I'd prefer to original way. WDYT ? @maropu

Comment (cloud-fan) [reply docs/sql-migration-guide.md]: AFAIK `numeric` is a data type defined by SQL that has precision and scale. This type is also implemented in SQL server and pgsql.

I think `number` is less confusing here.

**Comment by Ngone51** [comment]
Jenkins, retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: docs/sql-migration-guide.md (hunk 17 above)
Comment (maropu) [reply docs/sql-migration-guide.md]: Ur, I was mistaken. I meant number literals. Anyway, it doesn't matter. Thanks, guys.
